packages/core/src/modules/message-pìckup/MessagePickupApi.ts

```diff
@@ -0,0 +1,96 @@
+import type { MessagePickupProtocol } from './protocol/MessagePickupProtocol'
+import type { MessageRepository } from '../../storage/MessageRepository'
+
+import { AgentContext } from '../../agent'
+import { MessageSender } from '../../agent/MessageSender'
+import { OutboundMessageContext } from '../../agent/models'
+import { InjectionSymbols } from '../../constants'
+import { AriesFrameworkError } from '../../error'
+import { injectable } from '../../plugins'
+import { ConnectionService } from '../connections/services'
+
+import { MessagePickupModuleConfig } from './MessagePickupModuleConfig'
+
+export interface MessagePickupApi<MPPs extends MessagePickupProtocol[]> {
+  queueMessage(options: QueueMessageOptions): Promise<QueueMessageReturnType>
+  pickupMessages(options: PickupMessagesOptions<MPPs>): Promise<PickupMessagesReturnType>
+}
+
+@injectable()
+export class MessagePickupApi<MPPs extends MessagePickupProtocol[] = [V1MessagePickupProtocol, V2MessagePickupProtocol]>
+  implements MessagePickupApi<MPPs>
+{
+  public config: MessagePickupModuleConfig<MPPs>
+
+  private messageSender: MessageSender
+  private agentContext: AgentContext
+  private connectionService: ConnectionService
+
+  public constructor(
+    messageSender: MessageSender,
+    agentContext: AgentContext,
+    connectionService: ConnectionService,
+    config: MessagePickupModuleConfig<MPPs>
+  ) {
+    this.messageSender = messageSender
+    this.connectionService = connectionService
+    this.agentContext = agentContext
+    this.config = config
+  }
+
+  private getProtocol<MPP extends MPPs[number]['version']>(protocolVersion: MPP): MessagePickupProtocol {
+    const protocol = this.config.protocols.find((protocol) => protocol.version === protocolVersion)
+
+    if (!protocol) {
+      throw new AriesFrameworkError(`No message pickup protocol registered for protocol version ${protocolVersion}`)
+    }
+
+    return protocol
+  }
+
+  /**
+   * Add an encrypted message to the message pickup queue
+   *
+   * @param options: connectionId associated to the message and the encrypted message itself
+   */
+  public async queueMessage(options: QueueMessageOptions): Promise<QueueMessageReturnType> {
+    const connectionRecord = await this.connectionService.getById(this.agentContext, options.connectionId)
+
+    const messageRepository = this.agentContext.dependencyManager.resolve<MessageRepository>(
+      InjectionSymbols.MessageRepository
+    )
+
+    await messageRepository.add(connectionRecord.id, options.message)
+  }
+
+  /**
+   * Pickup queued messages from a message holder. It attempts to retrieve all current messages from the
+   * queue, receiving up to `batchSize` messages per batch retrieval.
+   *
+   * @param options connectionId, protocol version to use and batch size
+   */
+  public async pickupMessages(options: PickupMessagesOptions<MPPs>): Promise<PickupMessagesReturnType> {
```

Comment:
This doesn't actually wait for the pickup to be returned. Should we add this to function documentationion or something? 

If i understand correctly, mediation recipient will call the pickup in an interval, and you still need to handle the lifecycle in the mediation recipient module? Shouldn't we move all pickup logic, to the message pickup module? So picking up messages, whether implicit, explicit, live module, on an interval is managed by this module. The mediation recipient module is just a way to request mediation, register new keys, etc... (so really about the routing part)

Reply:
That's true: this `pickupMessages` will perform a single loop that can be called once or within an interval, and will return before the actual messages were pickup (e.g. Batch or Status/Delivery messages are received). Do you think it would be good to make it sync in order to block until the loop is completed or fail in the case of a timeout?

In this PR I left the lifecycle management in mediation module mostly to avoid introducing more breaking changes, but what I would like to do in a future is `MessagePickupApi` to create pickup loops that can be initiated/queried/stopped either by `MediationRecipientApi` or externally.

So I would say that MediationRecipientModule will be dedicated to the things you say and also holding the configuration (the `MediatorPickupStrategy`) and call `MessagePickupApi` to start/stop pickup loops as needed at the initialization/shutdown/mediator switching/etc.